Use global configs for keyframe interval

vigneshvg opened the pull request:
Instead of computing the required keyframe interval in libavif code and passing the "force keyframe" flag on individual frames, use each encoder's global configuration parameter to set the maximum keyframe interval.

For libaom: `kf_max_dist`.
For rav1e: `key_frame_interval`.
For svt-av1: `intra_period_length` and `force_key_frames`.

Here's the output obtained from various encoders for encoding an animated file with 10 frames with different values for --keyframe. Each line is read as follows:
--keyframe value: list of key frames before => list of keyframes after.

aom:
 --keyframe 0: same before and after.
 --keyframe 1: same before and after.
 --keyframe 2: 1,2,3,4,5,6,7,8,9,10 => 1,3,5,7,9 (the old behavior was not incorrect, but new one is exact).
 --keyframe 5: 1,2,6,7 =>1,6 (the old behavior was not incorrect, but new one is exact).

rav1e:
  * no changes. same output before and after.

svt-av1:
 --keyframe 0: same before and after.
 --keyframe 1: 1 => 1,2,3,4,5,6,7,8,9,10 (old behavior was incorrect).
 --keyframe 2: 1 => 1,3,5,7,9 (old behavior was incorrect).
 --keyframe 5: 1 => 1,6 (old behavior was incorrect).

To sum up:
 * svt-av1 was not producing the intended output with --keyframe before this change.
 * libaom had some weirdness in keyframe spacing when using the force keyframe flag on each frame and that weirdness goes away when using the kf_maxdist configuration (see: https://bugs.chromium.org/p/aomedia/issues/detail?id=3445).

The `AVIF_ADD_IMAGE_FLAG_FORCE_KEYFRAME` flag is retained in the codebase since there are some follow-up changes that will make use of the flag to synchronize keyframe placement between the color and alpha plane encoders (Issue #841).

wantehchang commented:
Vignesh: Thank you for the pull request. I will take a look tomorrow.

wantehchang commented on src/codec_rav1e.c:
```diff
@@ -172,6 +172,11 @@ static avifResult rav1eCodecEncodeImage(avifCodec * codec,
             }
         }
+        if (encoder->keyframeInterval > 0) {
+            if (rav1e_config_parse_int(rav1eConfig, "key_frame_interval", encoder->keyframeInterval) == -1) {
```
Nit: add a comment to document that "key_frame_interval" is the maximum interval between two keyframes.

vigneshvg replied:
Done.

wantehchang commented on src/codec_svt.c:
```diff
@@ -161,6 +161,15 @@ static avifResult svtCodecEncodeImage(avifCodec * codec,
+        if (encoder->keyframeInterval > 1) {
+            svt_config->intra_period_length = encoder->keyframeInterval - 1;
+        } else if (encoder->keyframeInterval == 1) {
+            // For this case, SV1-AV1 seems to need this flag set in svt_config
```
Optional: "Seems to" makes this comment sound speculative. It would be good to have an SVT-AV1 team member review this so we can write this comment with a definitive tone.

vigneshvg replied:
I have updated the comment to remove the speculation. But like you suggested, we can consult the SVT-AV1 team to make sure this is right.

wantehchang reviewed (changes requested):
In addition to the comments below, there are two typos in the commit message.

(1) In the following line:

```
 --keyframe 2: 1,2,3,4,5,6,7,8,9,19 => 1,3,5,7,9 (the old behavior was not incorrect, but new one is exact).
```

"19" seems to be a typo. Did you mean "10"?

(2) Change "kf_maxdist" to "kf_max_dist".

vigneshvg commented on src/codec_svt.c:
```diff
@@ -199,7 +208,7 @@ static avifResult svtCodecEncodeImage(avifCodec * codec,
 
     EbAv1PictureType frame_type = EB_AV1_INVALID_PICTURE;
-    if (addImageFlags & AVIF_ADD_IMAGE_FLAG_FORCE_KEYFRAME) {
+    if ((addImageFlags & AVIF_ADD_IMAGE_FLAG_FORCE_KEYFRAME) || (encoder->keyframeInterval == 1)) {
```
Ack.

wantehchang commented on src/codec_svt.c:
```diff
@@ -199,7 +208,7 @@ static avifResult svtCodecEncodeImage(avifCodec * codec,
 
     EbAv1PictureType frame_type = EB_AV1_INVALID_PICTURE;
-    if (addImageFlags & AVIF_ADD_IMAGE_FLAG_FORCE_KEYFRAME) {
+    if ((addImageFlags & AVIF_ADD_IMAGE_FLAG_FORCE_KEYFRAME) || (encoder->keyframeInterval == 1)) {
```
@hassount Hassene:

Could you review our changes in this file? We want to set SVT-AV1's equivalent of libaom's `kf_max_dist` to implement the maximum keyframe interval.

We also have a question: the libavif encoder has an `AVIF_ADD_IMAGE_FLAG_FORCE_KEYFRAME` flag to force a given animation frame to be a keyframe. We implemented that by setting `input_buffer->pic_type` to `EB_AV1_KEY_PICTURE` here, but that doesn't work. Do you know why?

hassount replied:
@wantehchang 
I am not sure I understand what we're trying to do with this:
>Could you review our changes in this file? We want to set SVT-AV1's equivalent of libaom's kf_max_dist to implement the maximum keyframe interval.

In general, SVT has only one parameter for the key-frame setting and lines 164-165 seem to be the right way of setting it. that said to set 
`else if (encoder->keyframeInterval == 1) {`
`svt_config->intra_period_length =` should be just set to `0`.

Now as for forcing the key frames, I believe you'd need to set both `force_key_frames` set to 1 (once for the sequence) and `input_buffer->pic_type to EB_AV1_KEY_PICTURE`

vigneshvg replied:
> @wantehchang I am not sure I understand what we're trying to do with this:
> 
> > Could you review our changes in this file? We want to set SVT-AV1's equivalent of libaom's kf_max_dist to implement the maximum keyframe interval.
> 
> In general, SVT has only one parameter for the key-frame setting and lines 164-165 seem to be the right way of setting it. that said to set `else if (encoder->keyframeInterval == 1) {` `svt_config->intra_period_length =` should be just set to `0`.
> 

i tried setting `svt_config->intra_period_length = 0` but it did not work as intended. (i.e.) the expectation is that all the frames will be keyframes in that case. but the output i got was the first frame was keyframes and subsequent frames were not.

any idea what i am missing?

> Now as for forcing the key frames, I believe you'd need to set both `force_key_frames` set to 1 (once for the sequence) and `input_buffer->pic_type to EB_AV1_KEY_PICTURE`

this is what this PR is now doing for the "all frames are key frames case".

hassount replied:
@1480c1 did we change the API to remove the +1 for intra_period ?
@vigneshvg can you try svt_config->intra_period_length = 1 instead ?

vigneshvg replied:
> @vigneshvg can you try svt_config->intra_period_length = 1 instead ?

yes, but that just outputs a keyframe every other frame. in this case, we want every frame to be a keyframe.

hassount replied:
this is weird then. either 0 or 1 should work @1480c1 any thoughts on this ?

wantehchang replied:
@hassount @1480c1 I filed https://gitlab.com/AOMediaCodec/SVT-AV1/-/issues/2076 about the intra_period_length == 0 issue discussed above.

vigneshvg commented on include/avif/avif.h:
```diff
@@ -1138,7 +1138,7 @@ typedef struct avifEncoder
     int maxThreads;
     int speed;
-    int keyframeInterval;     // How many frames between automatic forced keyframes; 0 to disable (default).
+    int keyframeInterval;     // Maximum interval between two keyframes; 0 to disable (default).
```
Done.

vigneshvg commented:
@wantehchang I have updated `codec_svt.c` to set force_key_frames unconditionally and documented why.

i have also updated the documentation in `avif.h` and `avifenc` to clealy define what we mean by "interval" (i.e.) the maximum allowed difference in frame indexes of any two subsequent keyframes.

Please take another look at this PR.

wantehchang commented on apps/avifenc.c:
```diff
@@ -111,7 +111,7 @@ static void syntax(void)
     printf("    --timescale,--fps V               : Set the timescale to V. If all frames are 1 timescale in length, this is equivalent to frames per second (Default: 30)\n");
     printf("                                        If neither duration nor timescale are set, avifenc will attempt to use the framerate stored in a y4m header, if present.\n");
-    printf("    -k,--keyframe INTERVAL            : Set the forced keyframe interval (maximum frames between keyframes). Set to 0 to disable (default).\n");
+    printf("    -k,--keyframe INTERVAL            : Set the maximum keyframe interval (maximum difference between the frame indexes of any two subsequent keyframes). Set to 0 to disable (default).\n");
```
Nit: "subsequent" doesn't sound right. Maybe "consecutive"?

("consecutive" may imply there are no frames in between, so it is also a little confusing.)

Perhaps we can go back to the previous description for now (you can find it in patchset 3 at https://aomedia-review.googlesource.com/c/libavif/+/175462) and then try to come up with a better description later.

vigneshvg replied:
I updated the wording based on your discussion with @joedrago.

i think the new one looks unambiguous.

wantehchang commented on src/codec_svt.c:
```diff
@@ -161,6 +161,17 @@ static avifResult svtCodecEncodeImage(avifCodec * codec,
 
+        // In order for SVT-AV1 to force keyframes by setting pic_type to
+        // EB_AV1_KEY_PICTURE on any frame, force_key_frames has to be set.
+        svt_config->force_key_frames = 1;
```
Nit/optional: 1 => TRUE

Here is the relevant code in SVT-AV1:

```
    /**
     * @brief Signal that force-key-frames is enabled.
     *
     */
    Bool force_key_frames;
```

and

```
typedef uint8_t Bool;
#define FALSE 0
#define TRUE 1
```

vigneshvg replied:
Done.

wantehchang commented on src/codec_svt.c:
```diff
@@ -161,6 +161,17 @@ static avifResult svtCodecEncodeImage(avifCodec * codec,
+        svt_config->force_key_frames = 1;
+
+        // keyframeInterval == 1 case is handled when encoding each frame by
+        // settings pic_type to EB_AV1_KEY_PICTURE. For keyframeInterval > 1,
```
settings => setting

Note: It would be good to mention that setting intra_period_length to 0 doesn't work, even though it should.

I found the following comment and code in SVT-AV1:

```
            // If the Intra period length is 0, then introduce an intra for every picture
            if (scs->static_config.intra_period_length == 0)
                pcs->cra_flag = TRUE;
```

vigneshvg replied:
Done.

wantehchang reviewed (approved):
LGTM.

vigneshvg commented:
@wantehchang i am submitting this PR based on your earlier approval. please let me know if you have any further comments and i'll address them in a follow-up.

wantehchang commented on apps/avifenc.c:
```diff
@@ -111,7 +111,7 @@ static void syntax(void)
     printf("    --timescale,--fps V               : Set the timescale to V. If all frames are 1 timescale in length, this is equivalent to frames per second (Default: 30)\n");
     printf("                                        If neither duration nor timescale are set, avifenc will attempt to use the framerate stored in a y4m header, if present.\n");
-    printf("    -k,--keyframe INTERVAL            : Set the forced keyframe interval (maximum frames between keyframes). Set to 0 to disable (default).\n");
+    printf("    -k,--keyframe INTERVAL            : Set the keyframe interval (Aany set of INTERVAL consecutive frames will have at least one keyframe). Set to 0 to disable (default).\n");
```
1. Aany => any
2. I think it is fine to have "maximum" before "keyframe interval"

wantehchang replied:
Fixed in https://github.com/AOMediaCodec/libavif/pull/1406.

wantehchang commented on src/codec_svt.c:
```diff
@@ -161,6 +161,18 @@ static avifResult svtCodecEncodeImage(avifCodec * codec,
+
+        // keyframeInterval == 1 case is handled when encoding each frame by
+        // setting pic_type to EB_AV1_KEY_PICTURE. For keyframeInterval > 1,
+        // set the intra_period_length. Eventhough setting intra_period_length
```
Eventhough => Even though

wantehchang replied:
Fixed in https://github.com/AOMediaCodec/libavif/pull/1406.

wantehchang reviewed (commented):
LGTM.
